core/src/main/scala/kafka/security/authorizer/AclAuthorizer.scala

```diff
@@ -130,6 +130,10 @@ class AclAuthorizer extends Authorizer with Logging {
 
   @volatile
   private var aclCache = new scala.collection.immutable.TreeMap[ResourcePattern, VersionedAcls]()(new ResourceOrdering)
+
+  private val resourceCache = new scala.collection.mutable.HashMap[ResourceIndex,
+    scala.collection.mutable.HashSet[String]]()
```

Comment:
Would scala "foreach" throw any exception when READ operation races with WRITE in HashMap / HashSet? If not, I think we can tolerate some READ inconsistency as ZK is also broadcasting the ACL changes asynchronously to brokers.

Reply:
I tested a bit, using 1 bg thread adding elements in order and removing those added in order while the main thread query the HashSet using "foreach". The elements in the query result are not in order, which means the concurrent R/W will destroy the "foreach" completely. 

Let me test the performance of the immutable map a bit. I prefer this approach since it won't lock the read operation, which is much more frequent.

Reply:
I tested a bit, using 1 bg thread adding and removing elements to a mutable.HashSet while the main thread constantly iterating the HashSet using "foreach". The "foreach" call doesn't throw any exception.

Reply:
Use Immutable collections:

Benchmark                               (aclCount)  (denyPercentage)  (resourceCount)  Mode  Cnt      Score      Error  Units
AclAuthorizerBenchmark.testUpdateCache          50               100           200000  avgt    5  10639.073 ± 3470.889  ms/op


Use Mutable collections:

AclAuthorizerBenchmark.testUpdateCache                      50               100           200000  avgt    5  4927.832 ± 2570.786  ms/op


The time cost doubled with immutable collections. But since adding 100 * 20000 ACL bindings only takes 10 seconds, adding 20000 ACL binding takes only 100 ms, which should be acceptable.

Reply:
Use Immutable collections:

Benchmark                                           (aclCount)  (denyPercentage)  (resourceCount)  Mode  Cnt     Score      Error  Units
AclAuthorizerBenchmark.testAclsIterator                     50               100           200000  avgt    5  4132.824 ± 2967.122  ms/op
AclAuthorizerBenchmark.testAuthorizeByResourceType          50               100           200000  avgt    5    46.733 ±    5.397  ms/op
AclAuthorizerBenchmark.testAuthorizer                       50               100           200000  avgt    5     6.844 ±    0.915  ms/op
AclAuthorizerBenchmark.testUpdateCache                      50               100           200000  avgt    5  7219.696 ± 4018.189  ms/op
JMH benchmarks done



Use Mutable collections:

AclAuthorizerBenchmark.testUpdateCache                      50               100           200000  avgt    5  4927.832 ± 2570.786  ms/op


The time cost is 2.3 seconds more. with immutable collections. But since adding 100 * 20000 ACL bindings only takes ~7 seconds, I think the performance should be acceptable.

Reply:
commit 62c44ade550a90671ff41bfb847e2bc28adc7baa